modules/atomic-widgets/base/atomic-widget-base.php

```diff
@@ -19,7 +19,60 @@ public function __construct( $data = [], $args = null ) {
 		$this->styles = $data['styles'] ?? [];
 	}
 
-	abstract public function get_atomic_controls(): array;
+	public function get_atomic_controls() {
+		$controls = $this->define_atomic_controls();
+
+		return $this->get_valid_controls( $controls );
+	}
+
+	private function get_valid_controls( array $controls ): array {
+		$valid_controls = [];
+		$schema = static::get_props_schema();
+
+		foreach ( $controls as $control ) {
+			if ( $control instanceof Section ) {
+				$cloned_section = clone $control;
+
+				$cloned_section->set_items(
+					$this->get_valid_controls( $control->get_items() )
+				);
+
+				$valid_controls[] = $cloned_section;
+				continue;
+			}
+
+			if ( ! ( $control instanceof Atomic_Control_Base ) ) {
+				$this->safe_throw( 'Control must be an instance of `Atomic_Control_Base`.' );
+				continue;
+			}
+
+			$prop_name = $control->get_bind();
+
+			if ( ! $prop_name ) {
+				$this->safe_throw( 'Control is missing a bound prop from the schema.' );
+				continue;
+			}
+
+			if ( ! array_key_exists( $prop_name, $schema ) ) {
+				$this->safe_throw( "Prop `{$prop_name}` is not defined in the schema of `{$this->get_name()}`. Did you forget to define it?" );
+				continue;
+			}
+
+			$valid_controls[] = $control;
+		}
+
+		return $valid_controls;
+	}
+
+	private function safe_throw( string $message ) {
+		if ( ! defined( 'ELEMENTOR_DEBUG' ) || ! ELEMENTOR_DEBUG ) {
+			return;
+		}
+
+		throw new \Exception( $message );
+	}
```

Comment:
Maybe create general util for other use cases and classes

Reply:
NO